node/wrapper-allocator/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::sync::atomic::{ Ordering::SeqCst, AtomicUsize };
+use core::alloc::{GlobalAlloc, Layout};
+use tikv_jemallocator::Jemalloc;
+
+pub struct WrapperAllocatorData {
+    allocated: AtomicUsize,
+    checkpoint: AtomicUsize,
+    peak: AtomicUsize,
+    // limit: AtomicUsize, // Should we introduce a checkpoint limit and fail allocation if the limit is hit?
+}
+
+impl WrapperAllocatorData {
+    /// Marks a new checkpoint. Returns peak allocation, in bytes, since the last checkpoint.
+    pub fn checkpoint(&self) -> usize {
+        let allocated = ALLOCATOR_DATA.allocated.load(SeqCst);
+        let old_cp = ALLOCATOR_DATA.checkpoint.swap(allocated, SeqCst);
+        ALLOCATOR_DATA.peak.swap(allocated, SeqCst).saturating_sub(old_cp)
+    }
+}
+
+pub static ALLOCATOR_DATA: WrapperAllocatorData = WrapperAllocatorData { allocated: AtomicUsize::new(0), checkpoint: AtomicUsize::new(0), peak: AtomicUsize::new(0) };
+
+struct WrapperAllocator<A: GlobalAlloc>(A);
+
+unsafe impl<A: GlobalAlloc> GlobalAlloc for WrapperAllocator<A> {
+
+    // SAFETY: The wrapped methods are as safe as the underlying allocator implementation is.
+
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let old_alloc = ALLOCATOR_DATA.allocated.fetch_add(layout.size(), SeqCst);
+        ALLOCATOR_DATA.peak.fetch_max(old_alloc + layout.size(), SeqCst);
+        self.0.alloc(layout)
```

Comment:
I think the memory ordering can just be `Relaxed` right now, since the atomics are being used as counters and not to synchronize other memory operations. `Relaxed` still gives you a total ordering between accesses to the same atomic.

Locks would make this slightly more deterministic but it doesn't seem worth the cost. Say you have a situation like thread1 calling `alloc` and getting `old_alloc`, thread2 stepping in and calling `dealloc`, and then thread1 setting the max `peak` to something higher than was actually allocated. You can prevent this with a lock but not totally solve it. There is still some indeterminacy where a worker may call the `dealloc` thread first and not hit the max at all while other workers would hit it. It's an edge case and not sure how feasible it would be to craft an attack around that.

But I think we decided on using `cgroups` to set the limit though which seems the way to go given the above. We can instead use this wrapper for detecting OOM by setting a flag before allocs and clearing it after.